File: Backend/src/main/java/com/sencarmarket/module/paiement/service/PaiementWebhookService.java
package com.sencarmarket.module.paiement.service;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.sencarmarket.module.paiement.entity.Paiement;
import com.sencarmarket.module.paiement.enums.StatutPaiement;
import com.sencarmarket.module.paiement.repository.PaiementRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.nio.charset.StandardCharsets;
import java.security.InvalidKeyException;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.time.LocalDateTime;
import java.util.Base64;
import java.util.List;
import java.util.Locale;

@Service
@RequiredArgsConstructor
@Slf4j
public class PaiementWebhookService {

    private static final String HMAC_SHA_256 = "HmacSHA256";

    private final PaiementRepository paiementRepository;
    private final ObjectMapper objectMapper;
    private final PaiementLogService paiementLogService;

    @Value("${paiements.wave.secret:}")
    private String waveSecret;

    @Value("${paiements.om.secret:}")
    private String omSecret;

    public String processWaveWebhook(String payload, String signature) {
        if (!verifyWebhookSignature(payload, signature, waveSecret)) {
            return "INVALID_SIGNATURE";
        }
        return processWebhookPayload(payload, "WAVE");
    }

    public String processOrangeMoneyWebhook(String payload, String signature) {
        if (!verifyWebhookSignature(payload, signature, omSecret)) {
            return "INVALID_SIGNATURE";
        }
        return processWebhookPayload(payload, "ORANGE_MONEY");
    }

    public boolean verifyWebhookSignature(String payload, String signature, String secret) {
        if (secret == null || secret.isBlank() || payload == null || signature == null || signature.isBlank()) {
            return false;
        }

        String normalizedSignature = normalizeSignature(signature);
        try {
            Mac mac = Mac.getInstance(HMAC_SHA_256);
            mac.init(new SecretKeySpec(secret.getBytes(StandardCharsets.UTF_8), HMAC_SHA_256));
            byte[] digest = mac.doFinal(payload.getBytes(StandardCharsets.UTF_8));

            String expectedHex = bytesToHex(digest);
            String expectedBase64 = Base64.getEncoder().encodeToString(digest);

            return MessageDigest.isEqual(normalizedSignature.getBytes(StandardCharsets.UTF_8),
                    expectedHex.getBytes(StandardCharsets.UTF_8))
                    || MessageDigest.isEqual(normalizedSignature.getBytes(StandardCharsets.UTF_8),
                    expectedBase64.getBytes(StandardCharsets.UTF_8));
        } catch (NoSuchAlgorithmException | InvalidKeyException e) {
            log.error("Erreur lors de la vérification de signature webhook", e);
            return false;
        }
    }

    private String processWebhookPayload(String payload, String provider) {
        try {
            JsonNode root = objectMapper.readTree(payload);
            String status = readFirstNonBlank(root,
                    "status", "event", "payment_status", "transaction_status", "etat", "state");
            String externalRef = readFirstNonBlank(root,
                    "reference_externe", "external_reference", "externalRef", "reference", "transaction_id", "id");
            String transactionRef = readFirstNonBlank(root,
                    "reference_transaction", "merchant_reference", "reference", "transaction_ref", "tx_ref");

            if (externalRef == null && transactionRef == null) {
                return "IGNORED";
            }

            Paiement paiement = findPaiementByWebhookReference(externalRef, transactionRef);
            if (paiement == null) {
                return "NOT_FOUND";
            }

            StatutPaiement nouveauStatut = mapWebhookStatus(status);
            if (nouveauStatut == null) {
                paiementLogService.createLogAction(paiement.getId(), "WEBHOOK_IGNORED",
                        String.format("Provider=%s, statut non mappé='%s'", provider, status));
                return "IGNORED";
            }

            applyWebhookStatus(paiement, nouveauStatut, externalRef, provider, status);
            return "SUCCESS";
        } catch (Exception e) {
            log.error("Erreur de parsing webhook {}: {}", provider, e.getMessage(), e);
            return "INVALID_PAYLOAD";
        }
    }

    private Paiement findPaiementByWebhookReference(String externalRef, String transactionRef) {
        if (externalRef != null) {
            List<Paiement> byExternal = paiementRepository.findByReferenceExterne(externalRef);
            if (!byExternal.isEmpty()) {
                return byExternal.get(0);
            }
        }
        if (transactionRef != null) {
            List<Paiement> byTransaction = paiementRepository.findByReferenceTransaction(transactionRef);
            if (!byTransaction.isEmpty()) {
                return byTransaction.get(0);
            }
        }
        return null;
    }

    private void applyWebhookStatus(Paiement paiement, StatutPaiement nouveauStatut, String externalRef,
                                    String provider, String rawStatus) {
        StatutPaiement ancien = paiement.getStatut();
        if (ancien == nouveauStatut && paiement.getDatePaiement() != null) {
            paiementLogService.createLogAction(paiement.getId(), "WEBHOOK_DUPLICATE",
                    String.format("Provider=%s, statut=%s", provider, rawStatus));
            return;
        }
        paiement.setStatut(nouveauStatut);
        if (externalRef != null && (paiement.getReferenceExterne() == null || paiement.getReferenceExterne().isBlank())) {
            paiement.setReferenceExterne(externalRef);
        }
        if (nouveauStatut == StatutPaiement.CONFIRME && paiement.getDatePaiement() == null) {
            paiement.setDatePaiement(LocalDateTime.now());
        }
        paiementRepository.save(paiement);
        paiementLogService.createLogAction(paiement.getId(), "WEBHOOK_UPDATE",
                String.format("Provider=%s, statut brut=%s, %s -> %s", provider, rawStatus, ancien, nouveauStatut));
    }

    private StatutPaiement mapWebhookStatus(String status) {
        if (status == null || status.isBlank()) {
            return null;
        }
        String normalized = status.trim().toUpperCase(Locale.ROOT);
        if (normalized.contains("SUCCESS") || normalized.contains("SUCCES")
                || normalized.contains("PAID") || normalized.contains("CONFIRM")) {
            return StatutPaiement.CONFIRME;
        }
        if (normalized.contains("FAIL") || normalized.contains("ERROR")
                || normalized.contains("ECHOUE") || normalized.contains("FAILED")) {
            return StatutPaiement.ECHOUE;
        }
        if (normalized.contains("CANCEL") || normalized.contains("ANNULE")) {
            return StatutPaiement.ANNULE;
        }
        if (normalized.contains("REFUND") || normalized.contains("REMBOUR")) {
            return StatutPaiement.REMBOURSE;
        }
        if (normalized.contains("PENDING") || normalized.contains("WAIT")
                || normalized.contains("EN_ATTENTE")) {
            return StatutPaiement.EN_ATTENTE;
        }
        return null;
    }

    private String readFirstNonBlank(JsonNode root, String... fields) {
        for (String field : fields) {
            JsonNode node = root.path(field);
            if (!node.isMissingNode() && !node.isNull()) {
                String value = node.asText(null);
                if (value != null && !value.isBlank()) {
                    return value.trim();
                }
            }
        }
        JsonNode dataNode = root.path("data");
        if (!dataNode.isMissingNode() && dataNode.isObject()) {
            for (String field : fields) {
                JsonNode node = dataNode.path(field);
                if (!node.isMissingNode() && !node.isNull()) {
                    String value = node.asText(null);
                    if (value != null && !value.isBlank()) {
                        return value.trim();
                    }
                }
            }
        }
        return null;
    }

    private String normalizeSignature(String signature) {
        String normalized = signature.trim();
        if (normalized.regionMatches(true, 0, "sha256=", 0, 7)) {
            normalized = normalized.substring(7);
        }
        return normalized;
    }

    private String bytesToHex(byte[] bytes) {
        StringBuilder builder = new StringBuilder(bytes.length * 2);
        for (byte b : bytes) {
            builder.append(String.format("%02x", b));
        }
        return builder.toString();
    }
}

File: Backend/src/modules/paiement/services/paiement-webhook.service.ts
import { createHmac, timingSafeEqual } from 'crypto';

import { Inject, Injectable } from '@nestjs/common';

import { PAIEMENT_REPOSITORY_PORT, PaiementRepositoryPort } from '../paiement.repository.port';
import type { StatutPaiement } from '../types/paiement.types';

import { PaiementLogService } from './paiement-log.service';

@Injectable()
export class PaiementWebhookService {
  constructor(
    @Inject(PAIEMENT_REPOSITORY_PORT) private readonly repository: PaiementRepositoryPort,
    private readonly paiementLogService: PaiementLogService,
  ) {}

  async processWebhook(payload: string, signature: string, secret: string, provider: string): Promise<string> {
    if (!this.verifyWebhookSignature(payload, signature, secret)) {
      return 'INVALID_SIGNATURE';
    }

    return this.processWebhookPayload(payload, provider);
  }

  private async processWebhookPayload(payload: string, provider: string): Promise<string> {
    try {
      const root = JSON.parse(payload) as Record<string, unknown>;
      const status = this.readFirstNonBlank(root, [
        'status',
        'event',
        'payment_status',
        'transaction_status',
        'etat',
        'state',
      ]);
      const externalRef = this.readFirstNonBlank(root, [
        'reference_externe',
        'external_reference',
        'externalRef',
        'reference',
        'transaction_id',
        'id',
      ]);
      const transactionRef = this.readFirstNonBlank(root, [
        'reference_transaction',
        'merchant_reference',
        'reference',
        'transaction_ref',
        'tx_ref',
      ]);

      if (!externalRef && !transactionRef) {
        return 'IGNORED';
      }

      const paiement = externalRef
        ? await this.repository.findPaiementByReferenceExterne(externalRef)
        : transactionRef
          ? await this.repository.findPaiementByReferenceTransaction(transactionRef)
          : null;

      if (!paiement) {
        return 'NOT_FOUND';
      }

      const nouveauStatut = this.mapWebhookStatus(status);
      if (!nouveauStatut) {
        await this.paiementLogService.createLogAction(
          paiement.id,
          'WEBHOOK_IGNORED',
          `Provider=${provider}, statut non mappé='${status ?? ''}'`,
        );
        return 'IGNORED';
      }

      if (paiement.statut === nouveauStatut && paiement.date_paiement) {
        await this.paiementLogService.createLogAction(
          paiement.id,
          'WEBHOOK_DUPLICATE',
          `Provider=${provider}, statut=${status ?? ''}`,
        );
        return 'SUCCESS';
      }

      await this.repository.updatePaiement(paiement.id, {
        statut: nouveauStatut,
        reference_externe: paiement.reference_externe ?? externalRef ?? undefined,
        date_paiement: nouveauStatut === 'CONFIRME' ? new Date() : undefined,
        updated_at: new Date(),
      });

      await this.paiementLogService.createLogAction(
        paiement.id,
        'WEBHOOK_UPDATE',
        `Provider=${provider}, statut brut=${status ?? ''}, ${paiement.statut ?? ''} -> ${nouveauStatut}`,
      );
      return 'SUCCESS';
    } catch {
      return 'INVALID_PAYLOAD';
    }
  }

  private verifyWebhookSignature(payload: string, signature: string, secret: string): boolean {
    if (!secret || !payload || !signature) {
      return false;
    }

    const normalized = signature.trim().replace(/^sha256=/i, '');
    const digest = createHmac('sha256', secret).update(payload, 'utf8').digest();
    const expectedHex = digest.toString('hex');
    const expectedBase64 = digest.toString('base64');

    return this.safeEqual(normalized, expectedHex) || this.safeEqual(normalized, expectedBase64);
  }

  private safeEqual(a: string, b: string): boolean {
    const aBuffer = Buffer.from(a, 'utf8');
    const bBuffer = Buffer.from(b, 'utf8');
    if (aBuffer.length !== bBuffer.length) {
      return false;
    }
    return timingSafeEqual(aBuffer, bBuffer);
  }

  private readFirstNonBlank(root: Record<string, unknown>, fields: string[]): string | null {
    for (const field of fields) {
      const value = root[field];
      if (typeof value === 'string' && value.trim().length > 0) {
        return value.trim();
      }
    }

    const data = root.data;
    if (data && typeof data === 'object' && !Array.isArray(data)) {
      const nested = data as Record<string, unknown>;
      for (const field of fields) {
        const value = nested[field];
        if (typeof value === 'string' && value.trim().length > 0) {
          return value.trim();
        }
      }
    }

    return null;
  }

  private mapWebhookStatus(status: string | null): StatutPaiement | null {
    if (!status) {
      return null;
    }

    const normalized = status.toUpperCase().trim();
    if (['SUCCESS', 'SUCCES', 'PAID', 'CONFIRM'].some((item) => normalized.includes(item))) {
      return 'CONFIRME';
    }
    if (['FAIL', 'ERROR', 'ECHOUE', 'FAILED'].some((item) => normalized.includes(item))) {
      return 'ECHOUE';
    }
    if (['CANCEL', 'ANNULE'].some((item) => normalized.includes(item))) {
      return 'ANNULE';
    }
    if (['REFUND', 'REMBOUR'].some((item) => normalized.includes(item))) {
      return 'REMBOURSE';
    }
    if (['PENDING', 'WAIT', 'EN_ATTENTE'].some((item) => normalized.includes(item))) {
      return 'EN_ATTENTE';
    }

    return null;
  }
}
